### PR TITLE
Added querySystemLimits for CSI spec 1.6

### PR DESCRIPTION
**Pull request by KshitijaKakde**
# Description
Added querySystemLimits for CSI spec 1.6

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/905

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [X] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
Tested the changes using integartion test and unit test

image

{"maxVal":"512","description":"Total Number of concurrent Move Pairs operations","type":"concurrentMovePairsOperations"},{"maxVal":"65536","minVal":"8","description":"Single replication volume size in GB","type":"replicationVolumeSizeGb"},{"maxVal":"262143","description":"Number of mappings between volumes and hosts","type":"replicationPairsPerReplicationConsistencyGroup"},{"maxVal":"512","description":"Number of host connections (IO controllers) per SDT","type":"ioControllersPerSdt"},{"maxVal":"128","description":"Depth of IO controller queue","type":"ioControllersQueueDepth"},{"maxVal":"32","description":"Number of IO controller queues","type":"ioControllersNumQueues"},{"maxVal":"8","description":"Number of paths per volume","type":"pathsPerVolume"},{"maxVal":"16","description":"Number of connections per host per protection domain","type":"connectionsPerHostPerProtectionDomain"},{"maxVal":"65519","description":"Number of host connections (IO controllers) per SYSTEM","type":"ioControllersPerSystem"}]}
    0
1048576
max volume size in volumeSizeGb 1048576

### CODE REVIEW

File: system_limit.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+package goscaleio
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	types "github.com/dell/goscaleio/types/v1"
+)
+
```
Comment (VamsiSiddu-7) [inline system_limit.go:22]: replace the name empty body with QuerySystemLimitsParam

File: system_limit.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+type emptyBody struct {
+}
+
+// GetSystemLimits gets list of sytem limits
+
+func (c *Client) GetSystemLimits() (syslimit *types.Limit, err error) {
+	defer TimeSpent("GetSystemLimits", time.Now())
+	var e emptyBody
```
Comment (VamsiSiddu-7) [inline system_limit.go:30]: same here instead of e give some meaningful variable name

**Review by VamsiSiddu-7 — CHANGES_REQUESTED** [review]
The variable names in go follow some specific naming conventions example systemLimit etc.

File: system_limit.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	path := "/api/instances/System/action/querySystemLimits"
+	err = c.getJSONWithRetry(
+		http.MethodPost, path, e, &syslimit)
+	if err != nil {
+		return nil, err
+	}
+
+	return syslimit, nil
```
Comment (VamsiSiddu-7) [inline system_limit.go:38]: systemLimits

File: system_limit.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// GetMaxVol returns max volume size in GB
+func (c *Client) GetMaxVol() (sys string, err error) {
+	defer TimeSpent("GetMaxVol", time.Now())
+	maxlimitType, err := c.GetSystemLimits()
```
Comment (VamsiSiddu-7) [inline system_limit.go:44]: systemLimits

File: types/v1/types.go [hunk 5]
```diff
@@ -1738,3 +1738,15 @@ type MDMQueueCommandDetails struct {
...
+
+// QuerySystemLimits defines struct for query system limits
+type QuerySystemLimits struct {
+	Type        string `json:"type,omitempty"`
+	Description string `json:"description,omitempty"`
+	MaxVal      string `json:"maxVal,omitempty"`
+}
+
+// SystemLimitEntryList defines struct for system limit entryList
+type Limit struct {
```
Comment (VamsiSiddu-7) [inline types/v1/types.go:1750]: I think it will be good to rename it to QuerySystemLimitsResponse

File: types/v1/types.go [hunk 6]
```diff
@@ -1738,3 +1738,15 @@ type MDMQueueCommandDetails struct {
 	TargetEntityIdentifier string    `json:"targetEntityIdentifier,omitempty"`
 	AllowedPhase           string    `json:"allowedPhase,omitempty"`
 }
+
+// QuerySystemLimits defines struct for query system limits
+type QuerySystemLimits struct {
```
Comment (VamsiSiddu-7) [inline types/v1/types.go:1743]: rename it to SystemLimit

File: system_limit.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	types "github.com/dell/goscaleio/types/v1"
+)
+
+type QuerySystemLimitsParam struct {
```
Comment (VamsiSiddu-7) [inline system_limit.go:23]: move this to types/v1 types.go because all the types are defined there only

**Review by VamsiSiddu-7 — CHANGES_REQUESTED** [review]

File: system_limit.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+		return nil, err
+	}
+
+	return systemLimits, nil
+}
+
+// GetMaxVol returns max volume size in GB
+func (c *Client) GetMaxVol() (systemLimits string, err error) {
+	defer TimeSpent("GetMaxVol", time.Now())
+	maxLimitType, err := c.GetSystemLimits()
```
Comment (VamsiSiddu-7) [inline system_limit.go:44]: change the variable name to systemLimits

File: system_limit.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+	if err != nil {
+		return "", err
+	}
+
+	for _, systemType := range maxLimitType.SystemLimitEntryList {
```
Comment (VamsiSiddu-7) [inline system_limit.go:50]: systemLimit

File: system_limit.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+		if systemType.Type == "volumeSizeGb" {
```
Comment (VamsiSiddu-7) [inline system_limit.go:52]: systemLimit.Type

File: system_limit.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+		http.MethodPost, path, body, &systemLimits)
+	if err != nil {
+		return nil, err
+	}
+
+	return systemLimits, nil
+}
+
+// GetMaxVol returns max volume size in GB
+func (c *Client) GetMaxVol() (systemLimits string, err error) {
```
Comment (VamsiSiddu-7) [inline system_limit.go:38]: This should return MaxVolumeSize not systemLimits change it.

**Review by VamsiSiddu-7 — APPROVED** [review]

**Comment by VamsiSiddu-7** [comment]
please re run the integration and unit tests to make sure nothing is breaking

**Review by suryagupta4 — APPROVED** [review]

**Pull request merged**